spec/compiler/semantic/case_spec.cr

```diff
@@ -0,0 +1,237 @@
+  it "checks exhaustiveness of union type" do
+    assert_warning %(
+        a = 1 || nil
+        case a
+        when Int32
+        end
+      ),
+      "warning in line 4\nWarning: case is not exhaustive.\n\nMissing types: Nil"
+  end
+
+  it "checks exhaustiveness of single type" do
+    assert_warning %(
+        case 1
+        when Nil
+        end
+      ),
+      "warning in line 3\nWarning: case is not exhaustive.\n\nMissing types: Int32"
+  end
+
+  it "covers all types" do
+    assert_no_warnings %(
+        a = 1 || nil
+        case a
+        when Int32
+        when Nil
+        end
+      )
+  end
+
+  it "can't prove exhaustiveness" do
+    assert_warning %(
+        struct Int32
+          def ===(other)
+            true
+          end
+        end
+
+        case 1
+        when 2
+        end
+      ),
+      "warning in line 9\nWarning: can't prove case is exhaustive.\n\nPlease add an `else` clause."
+  end
+
+  it "checks exhaustiveness of bool type (missing true)" do
+    assert_warning %(
+        struct Bool
+          def ===(other)
+            true
+          end
+        end
+
+        case false
+        when false
+        end
+      ),
+      "warning in line 9\nWarning: case is not exhaustive.\n\nMissing cases: true"
+  end
+
+  it "checks exhaustiveness of bool type (missing false)" do
+    assert_warning %(
+        struct Bool
+          def ===(other)
+            true
+          end
+        end
+
+        case false
+        when true
+        end
+      ),
+      "warning in line 9\nWarning: case is not exhaustive.\n\nMissing cases: false"
+  end
+
+  it "checks exhaustiveness of enum via question method" do
+    assert_warning %(
+        struct Enum
+          def ==(other : self)
+            value == other.value
+          end
+        end
+
+        enum Color
+          Red
+          Green
+          Blue
+        end
+
+        e = Color::Red
+        case e
+        when .red?
+        end
+      ),
+      "warning in line 16\nWarning: case is not exhaustive for enum Color.\n\nMissing members: Green, Blue"
+  end
+
+  it "checks exhaustiveness of enum via const" do
+    assert_warning %(
+        struct Enum
+          def ==(other : self)
+            value == other.value
+          end
+
+          def ===(other)
+            true
+          end
+        end
+
+        enum Color
+          Red
+          Green
+          Blue
+        end
+
+        e = Color::Red
+        case e
+        when Color::Red
+        end
+      ),
+      "warning in line 20\nWarning: case is not exhaustive for enum Color.\n\nMissing members: Green, Blue"
+  end
+
+  it "checks exhaustiveness of enum (all cases covered)" do
+    assert_no_warnings %(
+        struct Enum
+          def ==(other : self)
+            value == other.value
+          end
+        end
+
+        enum Color
+          Red
+          Green
+          Blue
+        end
+
+        e = Color::Red
+        case e
+        when .red?
```

Comment:
Yes, it's not handled because an enum doesn't match `:red`. Symbols and enums are not comparable. Like, if you do that right now it will never match.

We _could_ make it work in the future (I can work on it before 1.0), but then if it works in `case` but it doesn't work in `==` or `===` it will be quite confusing, I think.

The way to go with this is:
- remove Symbol from the language
- then `:symbol` can only mean a shortcut for an enum, and then we can use that in exhaustive checks, equality, etc. (with a bit of effort)

But for now exhaustiveness for enums is only supported by actual constants of by `.question?` method.

Reply:
(That said, we don't need to do any of that before 1.0, I would be fine with using just the question syntax, like what you have to do now even before this PR)

Reply:
I'm ok if `when :symbol` is not supported, i wanted to clarify. I forgot about Symbol/Enum not having `===`. I find it a bit odd that we are pushing `:value` on method calls due to autocasting, but we can't push it here; even if we know the type of the condition is a specific enum type.

It can be iterated later I guess.

Reply:
Yeah, it's a situation I don't like either.

I created https://github.com/crystal-lang/crystal/issues/8929 because of that. If we implement that, we can allow symbols in `case` when matching against enums too, and consider them for exhaustiveness.